File: src/Java/234.ListIsPalindrome.java
import java.awt.List;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

public class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode node = head;
        ListNode pre = null;
        while (node != null) {
            ListNode next = node.next;
            node.next = pre;
            pre = node;
            node = next;
        }
        return pre;
    }

    public ListNode findMiddle(ListNode head){
        ListNode slow = head;
        ListNode fast = head.next;
        while (fast != null && fast.next != null)
        {
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }

    public boolean isPalindrome(ListNode head) {
        if (head == null) {
            return true;
        }
        ListNode middle = findMiddle(head);
        ListNode tail = middle.next;
        middle.next = null;
        ListNode p = head;
        ListNode second = reverseList(tail);
        ListNode q = second;
        while (p != null && q != null){
            if (p.val != q.val)
                return false;
            p = p.next;
            q = q.next;
        }
        return true;
    }
}

File: src/Python/234.ListIsPalindrome.py
# Definition for singly-linked list.

from copy import deepcopy

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        stack = []
        node = head
        count = 0
        while node != None:
            stack.append(node.val)
            node = node.next
            count += 1
        for i in range(count):
            if stack[i] != stack[count - i - 1]:
                return False
        return True

if __name__ == '__main__':
    solution = Solution()
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(2)
    head.next.next.next = ListNode(3)
    print(solution.isPalindrome(head))
else:
    pass
